# Import libraries

In [1]:
import os
import pandas as pd
import numpy as np

In [2]:
from pybaseball import pitching_stats

In [3]:
# Batting Stats , qual is the number of players appearences 
pitching = pitching_stats(2007,2022, qual=200)
pitching

IDfg  Season              Name Team  Age   W   L  WAR   ERA   G  ...  \
1    10954    2018      Jacob deGrom  NYM   30  10   9  9.0  1.70  32  ...   
12    1303    2011      Roy Halladay  PHI   34  19   6  8.7  2.35  32  ...   
6     1943    2009      Zack Greinke  KCR   25  16   8  8.7  2.16  33  ...   
4     2036    2015   Clayton Kershaw  LAD   27  16   7  8.6  2.13  33  ...   
203   8700    2009  Justin Verlander  DET   26  19   9  8.4  3.45  35  ...   
..     ...     ...               ...  ...  ...  ..  ..  ...   ...  ..  ...   
327   2072    2013    Jeremy Guthrie  KCR   34  15  12  0.6  4.04  33  ...   
321   3551    2012   Clayton Richard  SDP   28  14  14  0.4  3.99  33  ...   
258   4366    2011      Joe Saunders  ARI   30  12  13  0.3  3.69  33  ...   
384   1116    2007   Livan Hernandez  ARI   32  11  11  0.2  4.93  33  ...   
386    150    2010     Rodrigo Lopez  ARI   34   7  16 -0.1  5.00  33  ...   

       LA  Barrels  Barrel%  maxEV  HardHit  HardHit%  Events  CStr%   CSW%  \
1    11.1       20    0.039  112.9      148     0.287     515  0.164  0.315   
12    NaN        0      NaN    NaN        0       NaN       0  0.185  0.294   
6     NaN        0      NaN    NaN        0       NaN       0  0.187  0.287   
4     8.9       15    0.028  113.3      132     0.244     542  0.164  0.323   
203   NaN        0      NaN    NaN        0       NaN       0  0.175  0.289   
..    ...      ...      ...    ...      ...       ...     ...    ...    ...   
327   NaN        0      NaN    NaN        0       NaN       0  0.185  0.236   
321   NaN        0      NaN    NaN        0       NaN       0  0.183  0.246   
258   NaN        0      NaN    NaN        0       NaN       0  0.169  0.231   
384   NaN        0      NaN    NaN        0       NaN       0  0.185  0.239   
386   NaN        0      NaN    NaN        0       NaN       0  0.199  0.250   

     xERA  
1     NaN  
12    NaN  
6     NaN  
4     NaN  
203   NaN  
..    ...  
327   NaN  
321   NaN  
258   NaN  
384   NaN  
386   NaN  

[391 rows x 334 columns]

In [5]:
# Export DataFrame to csv
pitching.to_csv('../static/data/Raw_Data/pitching.csv', index=False)

In [6]:
# Group by player ID and drop players with only 1 season of data 
# Leave players with a t least 2 seasons of data
pitching = pitching.groupby("IDfg", group_keys=False).filter(lambda x: x.shape[0]>1)

In [7]:
pitching

IDfg  Season              Name Team  Age   W   L  WAR   ERA   G  ...  \
1    10954    2018      Jacob deGrom  NYM   30  10   9  9.0  1.70  32  ...   
12    1303    2011      Roy Halladay  PHI   34  19   6  8.7  2.35  32  ...   
6     1943    2009      Zack Greinke  KCR   25  16   8  8.7  2.16  33  ...   
4     2036    2015   Clayton Kershaw  LAD   27  16   7  8.6  2.13  33  ...   
203   8700    2009  Justin Verlander  DET   26  19   9  8.4  3.45  35  ...   
..     ...     ...               ...  ...  ...  ..  ..  ...   ...  ..  ...   
343    976    2010        Randy Wolf  MIL   33  13  12  0.8  4.17  34  ...   
327   2072    2013    Jeremy Guthrie  KCR   34  15  12  0.6  4.04  33  ...   
321   3551    2012   Clayton Richard  SDP   28  14  14  0.4  3.99  33  ...   
258   4366    2011      Joe Saunders  ARI   30  12  13  0.3  3.69  33  ...   
384   1116    2007   Livan Hernandez  ARI   32  11  11  0.2  4.93  33  ...   

       LA  Barrels  Barrel%  maxEV  HardHit  HardHit%  Events  CStr%   CSW%  \
1    11.1       20    0.039  112.9      148     0.287     515  0.164  0.315   
12    NaN        0      NaN    NaN        0       NaN       0  0.185  0.294   
6     NaN        0      NaN    NaN        0       NaN       0  0.187  0.287   
4     8.9       15    0.028  113.3      132     0.244     542  0.164  0.323   
203   NaN        0      NaN    NaN        0       NaN       0  0.175  0.289   
..    ...      ...      ...    ...      ...       ...     ...    ...    ...   
343   NaN        0      NaN    NaN        0       NaN       0  0.181  0.248   
327   NaN        0      NaN    NaN        0       NaN       0  0.185  0.236   
321   NaN        0      NaN    NaN        0       NaN       0  0.183  0.246   
258   NaN        0      NaN    NaN        0       NaN       0  0.169  0.231   
384   NaN        0      NaN    NaN        0       NaN       0  0.185  0.239   

     xERA  
1     NaN  
12    NaN  
6     NaN  
4     NaN  
203   NaN  
..    ...  
343   NaN  
327   NaN  
321   NaN  
258   NaN  
384   NaN  

[326 rows x 334 columns]

In [8]:
# Write a function that takes WAR number from a player's season to insert in Next_WAR column
def next_season(player):
    player = player.sort_values("Season")
    player["Next_WAR"] = player["WAR"].shift(-1)
    return player

pitching = pitching.groupby("IDfg", group_keys=False).apply(next_season)

In [9]:
# Show the next WAR of each player in a DF 
# Missing values are for some seasons with no data (some players could not play in these seasons)
war_df=pitching[['Name','Season','WAR','Next_WAR']]
war_df

Name  Season  WAR  Next_WAR
241     Mark Buehrle    2007  3.3       4.2
286     Mark Buehrle    2008  4.2       3.0
295     Mark Buehrle    2009  3.0       3.8
353     Mark Buehrle    2010  3.8       3.1
237     Mark Buehrle    2011  3.1       1.6
..               ...     ...  ...       ...
160       Aaron Nola    2022  6.3       NaN
139  Sandy Alcantara    2021  4.3       5.7
11   Sandy Alcantara    2022  5.7       NaN
164     Shane Bieber    2019  5.5       4.9
78      Shane Bieber    2022  4.9       NaN

[326 rows x 4 columns]

## Cleaning Data

In [10]:
# Check null values
null_data=pitching.isnull().sum()
null_data

IDfg          0
Season        0
Name          0
Team          0
Age           0
           ... 
Events        0
CStr%         0
CSW%          0
xERA        326
Next_WAR     91
Length: 335, dtype: int64

In [11]:
# List columns with no missing values 
complete_cols=list(pitching.columns[null_data == 0])

In [12]:
# Create a clean DataFrame with complete columns and Next_WAR column
pitching = pitching[complete_cols + ["Next_WAR"]].copy()
pitching

IDfg  Season             Name Team  Age   W   L  WAR   ERA   G  ...  \
241    225    2007     Mark Buehrle  CHW   28  10   9  3.3  3.63  30  ...   
286    225    2008     Mark Buehrle  CHW   29  15  12  4.2  3.79  34  ...   
295    225    2009     Mark Buehrle  CHW   30  13  10  3.0  3.84  33  ...   
353    225    2010     Mark Buehrle  CHW   31  13  13  3.8  4.28  33  ...   
237    225    2011     Mark Buehrle  CHW   32  13   9  3.1  3.59  31  ...   
..     ...     ...              ...  ...  ...  ..  ..  ...   ...  ..  ...   
160  16149    2022       Aaron Nola  PHI   29  11  13  6.3  3.25  32  ...   
139  18684    2021  Sandy Alcantara  MIA   25   9  15  4.3  3.19  33  ...   
11   18684    2022  Sandy Alcantara  MIA   26  14   9  5.7  2.28  32  ...   
164  19427    2019     Shane Bieber  CLE   24  15   8  5.5  3.28  34  ...   
78   19427    2022     Shane Bieber  CLE   27  13   8  4.9  2.88  31  ...   

     Oppo%+  Soft%+  Med%+  Hard%+  Barrels  HardHit  Events  CStr%   CSW%  \
241      81     111    100      94        0        0       0  0.171  0.250   
286      83     115    100      91        0        0       0  0.171  0.239   
295      95      98    109      81        0        0       0  0.176  0.242   
353     100     124     99      87        0        0       0  0.181  0.242   
237      85     115     95      97        0        0       0  0.179  0.244   
..      ...     ...    ...     ...      ...      ...     ...    ...    ...   
160     104     103    100      98       38      169     534  0.197  0.324   
139      91     116    100      91       35      227     576  0.146  0.279   
11       93     126     99      87       32      239     620  0.153  0.275   
164     104      79     97     112       41      231     554  0.185  0.325   
78       92      74    102     111       40      238     555  0.175  0.314   

     Next_WAR  
241       4.2  
286       3.0  
295       3.8  
353       3.1  
237       1.6  
..        ...  
160       NaN  
139       5.7  
11        NaN  
164       4.9  
78        NaN  

[326 rows x 155 columns]

### Drop and Dummify all of the categorical variables 

In [13]:
#Check columns types 
pitching.dtypes

IDfg          int64
Season        int64
Name         object
Team         object
Age           int64
             ...   
HardHit       int64
Events        int64
CStr%       float64
CSW%        float64
Next_WAR    float64
Length: 155, dtype: object

In [14]:
# Find columns with object data type
pitching.dtypes[pitching.dtypes == 'object']

Name       object
Team       object
Dollars    object
Age Rng    object
dtype: object

In [15]:
# Drop 'Dollars' (Dollar value of player) and 'Age Rng' (Player's age range during a season) 
pitching = pitching.drop(['Dollars','Age Rng'], axis = 1, inplace = False)

In [16]:
# Assign each team name to number 
pitching["Team_code"] = pitching["Team"].astype("category").cat.codes

In [17]:
# Copy pitching data 
pitching_full = pitching.copy()

# Drop Nan Values from 'Next_Raw'
pitching = pitching.dropna().copy()

In [18]:
pitching

IDfg  Season             Name Team  Age   W   L  WAR   ERA   G  ...  \
241    225    2007     Mark Buehrle  CHW   28  10   9  3.3  3.63  30  ...   
286    225    2008     Mark Buehrle  CHW   29  15  12  4.2  3.79  34  ...   
295    225    2009     Mark Buehrle  CHW   30  13  10  3.0  3.84  33  ...   
353    225    2010     Mark Buehrle  CHW   31  13  13  3.8  4.28  33  ...   
237    225    2011     Mark Buehrle  CHW   32  13   9  3.1  3.59  31  ...   
..     ...     ...              ...  ...  ...  ..  ..  ...   ...  ..  ...   
360  13431    2016   Marcus Stroman  TOR   25   9  10  3.3  4.37  32  ...   
13   16149    2018       Aaron Nola  PHI   25  17   6  5.5  2.37  33  ...   
301  16149    2019       Aaron Nola  PHI   26  12   7  3.4  3.87  34  ...   
139  18684    2021  Sandy Alcantara  MIA   25   9  15  4.3  3.19  33  ...   
164  19427    2019     Shane Bieber  CLE   24  15   8  5.5  3.28  34  ...   

     Soft%+  Med%+  Hard%+  Barrels  HardHit  Events  CStr%   CSW%  Next_WAR  \
241     111    100      94        0        0       0  0.171  0.250       4.2   
286     115    100      91        0        0       0  0.171  0.239       3.0   
295      98    109      81        0        0       0  0.176  0.242       3.8   
353     124     99      87        0        0       0  0.181  0.242       3.1   
237     115     95      97        0        0       0  0.179  0.244       1.6   
..      ...    ...     ...      ...      ...     ...    ...    ...       ...   
360      98    100     101       28      264     631  0.172  0.264       3.4   
13      122    113      71       21      168     542  0.198  0.322       3.4   
301      92     95     110       34      210     532  0.206  0.315       6.3   
139     116    100      91       35      227     576  0.146  0.279       5.7   
164      79     97     112       41      231     554  0.185  0.325       4.9   

     Team_code  
241          6  
286          6  
295          6  
353          6  
237          6  
..         ...  
360         31  
13          22  
301         22  
139         16  
164          8  

[235 rows x 154 columns]

## Machine Learning Model to predict next WAR

In [19]:
from sklearn.linear_model import Ridge
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.model_selection import TimeSeriesSplit

# Initialize Ridge Regression Model
# If Lambda or alpa in Python is higher it reduces overfitting, if it's lower it's closer to regular linear regression
rr = Ridge(alpha=1)


# Split Data into 3 parts 
split = TimeSeriesSplit(n_splits=3)

# Go over the features to find the best one until it selects 20 
sfs = SequentialFeatureSelector(rr, 
                                n_features_to_select=20, 
                                direction="forward",
                                cv=split,
                                n_jobs=8
                               )

In [20]:
# Remove some columns 
removed_columns = ["Next_WAR", "Name", "Team", "IDfg", "Season"]
selected_columns = pitching.columns[~pitching.columns.isin(removed_columns)]

In [21]:
# All columns except the one we have removed 
selected_columns

Index(['Age', 'W', 'L', 'WAR', 'ERA', 'G', 'GS', 'CG', 'ShO', 'SV',
       ...
       'Oppo%+', 'Soft%+', 'Med%+', 'Hard%+', 'Barrels', 'HardHit', 'Events',
       'CStr%', 'CSW%', 'Team_code'],
      dtype='object', length=149)

In [22]:
X=pitching[selected_columns]
y=pitching['Next_WAR']

In [23]:
# Scale the data 
# MinMax Scales values to be between 0 and 1
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
pitching.loc[:,selected_columns] = scaler.fit_transform(X)

In [24]:
# Most values are now between 0 and 1
pitching.describe()

IDfg       Season         Age           W           L  \
count    235.000000   235.000000  235.000000  235.000000  235.000000   
mean    4383.293617  2011.459574    0.402128    0.472813    0.423404   
std     3694.105282     3.121723    0.190267    0.189575    0.183224   
min      225.000000  2007.000000    0.000000    0.000000    0.000000   
25%     1571.500000  2009.000000    0.277778    0.333333    0.312500   
50%     3254.000000  2011.000000    0.388889    0.500000    0.437500   
75%     6345.000000  2014.000000    0.555556    0.611111    0.562500   
max    19427.000000  2021.000000    1.000000    1.000000    1.000000   

              WAR         ERA           G          GS          CG  ...  \
count  235.000000  235.000000  235.000000  235.000000  235.000000  ...   
mean     0.466296    0.476003    0.594681    0.668085    0.188008  ...   
std      0.182282    0.177600    0.145530    0.162674    0.180151  ...   
min      0.000000    0.000000    0.000000    0.000000    0.000000  ...   
25%      0.352273    0.362216    0.500000    0.571429    0.090909  ...   
50%      0.465909    0.471591    0.625000    0.714286    0.090909  ...   
75%      0.579545    0.589489    0.750000    0.714286    0.272727  ...   
max      1.000000    1.000000    1.000000    1.000000    1.000000  ...   

           Soft%+       Med%+      Hard%+     Barrels     HardHit      Events  \
count  235.000000  235.000000  235.000000  235.000000  235.000000  235.000000   
mean     0.529730    0.539116    0.502837    0.116538    0.123694    0.144216   
std      0.175608    0.178639    0.172765    0.261058    0.273323    0.316218   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.391892    0.423077    0.398148    0.000000    0.000000    0.000000   
50%      0.513514    0.538462    0.500000    0.000000    0.000000    0.000000   
75%      0.662162    0.653846    0.592593    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

            CStr%        CSW%    Next_WAR   Team_code  
count  235.000000  235.000000  235.000000  235.000000  
mean     0.504945    0.476498    4.164681    0.468351  
std      0.184939    0.163777    1.804975    0.314981  
min      0.000000    0.000000    0.300000    0.000000  
25%      0.378378    0.369231    2.800000    0.187500  
50%      0.486486    0.476923    4.100000    0.437500  
75%      0.621622    0.561538    5.400000    0.781250  
max      1.000000    1.000000    9.000000    1.000000  

[8 rows x 152 columns]

In [25]:
# Fit the data 
sfs.fit(X, y)

SequentialFeatureSelector(cv=TimeSeriesSplit(gap=0, max_train_size=None, n_splits=3, test_size=None),
                          estimator=Ridge(alpha=1), n_features_to_select=20,
                          n_jobs=8)

In [26]:
# sfs.get_support() returns True to each column selected 
sfs.get_support()

array([False, False,  True,  True, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False,  True, False, False, False, False,
       False, False, False, False,  True, False,  True, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False, False,  True,  True, False,
        True, False, False, False, False,  True, False, False,  True,
       False, False,  True, False, False, False, False, False, False,
       False,  True, False, False, False, False, False,  True, False,
       False, False, False,  True, False, False, False, False, False,
       False, False, False,  True, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [27]:
# Assign sfs to a variable called predictors
# These are the 20 columns selected by our model
predictors = list(selected_columns[sfs.get_support()])
predictors

['L',
 'WAR',
 'G',
 'IBB',
 'IFH',
 'WHIP',
 'LOB%',
 'inLI',
 'FB% 2',
 'FBv',
 'wFB/C',
 'Z-Contact%',
 'F-Strike%',
 'SD',
 'RS/9',
 'Z-Contact% (sc)',
 'BIP-Wins',
 'Hard%',
 'LD%+',
 'Cent%+']

In [28]:
# We only want to use past data to predict future data 

def backtest(data, model, predictors, start=1, step=1):
    # Create a list to hold all seasons predictions
    all_predictions = []
    
    # List of seasons sorted by order 
    years = sorted(data["Season"].unique())
    
    for i in range(start, len(years), step):
        current_year = years[i]
        
        # Train set is all years before current_year
        train = data[data["Season"] < current_year]
        
        # Test set is the current_year
        test = data[data["Season"] == current_year]
        
        model.fit(train[predictors], train["Next_WAR"])
        
        # Generate predictions on test set
        preds = model.predict(test[predictors])
        preds = pd.Series(preds, index=test.index)
        
        # Concat predictions with actual 
        combined = pd.concat([test["Next_WAR"], preds], axis=1)
        combined.columns = ["actual", "prediction"]
        
        all_predictions.append(combined)
    return pd.concat(all_predictions)

In [29]:
model_pred=backtest(pitching, rr, predictors)
model_pred

actual  prediction
286     3.0    3.649569
46      5.8    5.271326
333     2.8    4.025636
96      3.6    3.909007
229     4.4    3.313639
..      ...         ...
13      3.4    5.142955
30      3.3    7.700713
301     6.3    4.012145
164     4.9    5.064353
139     5.7    4.948089

[211 rows x 2 columns]

## Caculate the accuracy

In [30]:
# Calculate the mean squared error value (To compare the prediction from actual value)
from sklearn.metrics import mean_squared_error

mean_squared_error(model_pred["actual"], model_pred["prediction"])

2.4442807018278527

In [31]:
# Square root of mean_squared_error
2.4442807018278527 ** 0.5

1.5634195539994544

In [32]:
pitching['Next_WAR'].describe()

count    235.000000
mean       4.164681
std        1.804975
min        0.300000
25%        2.800000
50%        4.100000
75%        5.400000
max        9.000000
Name: Next_WAR, dtype: float64

std = 1.804975

Square root of mean_squared_error = 1.5634195539994544 < std 
Which is good for the model

### Add player history data to improve accuracy
### Find correlation between WAR nb and season number for all previous season for player of IDf=225

In [33]:
MarkB=pitching[pitching["IDfg"]==225].copy()
MarkB["player_season"]=range(0,MarkB.shape[0])
MarkB[["player_season","WAR"]]

player_season       WAR
241              0  0.352273
286              1  0.454545
295              2  0.318182
353              3  0.409091
237              4  0.329545
271              5  0.159091
340              6  0.227273

In [34]:
MarkB[["player_season","WAR"]].expanding().corr()

player_season       WAR
241 player_season            NaN       NaN
    WAR                      NaN       NaN
286 player_season       1.000000  1.000000
    WAR                 1.000000  1.000000
295 player_season       1.000000 -0.240192
    WAR                -0.240192  1.000000
353 player_season       1.000000  0.072868
    WAR                 0.072868  1.000000
237 player_season       1.000000 -0.249513
    WAR                -0.249513  1.000000
271 player_season       1.000000 -0.659578
    WAR                -0.659578  1.000000
340 player_season       1.000000 -0.726517
    WAR                -0.726517  1.000000

In [35]:
# Slice the WAR nb from 1st row
MarkB[["player_season","WAR"]].expanding().corr().loc[(slice(None),"player_season"),"WAR"]

241  player_season         NaN
286  player_season    1.000000
295  player_season   -0.240192
353  player_season    0.072868
237  player_season   -0.249513
271  player_season   -0.659578
340  player_season   -0.726517
Name: WAR, dtype: float64

In [36]:
# Turn the correlations between WAR and the season number into list
list(MarkB[["player_season","WAR"]].expanding().corr().loc[(slice(None),"player_season"),"WAR"])

[nan,
 0.9999999999999998,
 -0.24019223070763157,
 0.07286792513358728,
 -0.2495131446207222,
 -0.6595780050719466,
 -0.726517479565979]

###  Function to find correlation between WAR nb and season nb of all previous seasons for each player

In [37]:
def player_history(df):
    df = df.sort_values("Season")
        
    df["player_season"] = range(0, df.shape[0])
    
    # Find the corr between "player_season" and "WAR"
    df["war_corr"] = list(df[["player_season", "WAR"]].expanding().corr().loc[(slice(None), "player_season"),"WAR"])
    
    df["war_corr"].fillna(0, inplace=True)
    
    # Divide WAR of current season by WAR of previous season
    df["war_diff"] = df["WAR"] / df["WAR"].shift(1)
    # Replace NaN values with 1 when there's no previous season 
    df["war_diff"].fillna(1, inplace=True)
    # Replace infinite value with 1
    df["war_diff"][df["war_diff"] == np.inf] = 1
    
    return df

pitching = pitching.groupby("IDfg", group_keys=False).apply(player_history)


In [38]:
# Find averages across seasons; tell us if a player performed better than the average of a season or no 
def group_averages(df):
    return df["WAR"] / df["WAR"].mean()

In [39]:
pitching["war_season"] = pitching.groupby("Season", group_keys=False).apply(group_averages)
pitching

IDfg  Season             Name Team       Age         W       L  \
241    225    2007     Mark Buehrle  CHW  0.388889  0.222222  0.3750   
286    225    2008     Mark Buehrle  CHW  0.444444  0.500000  0.5625   
295    225    2009     Mark Buehrle  CHW  0.500000  0.388889  0.4375   
353    225    2010     Mark Buehrle  CHW  0.555556  0.388889  0.6250   
237    225    2011     Mark Buehrle  CHW  0.611111  0.388889  0.3750   
..     ...     ...              ...  ...       ...       ...     ...   
360  13431    2016   Marcus Stroman  TOR  0.222222  0.166667  0.4375   
13   16149    2018       Aaron Nola  PHI  0.222222  0.611111  0.1875   
301  16149    2019       Aaron Nola  PHI  0.277778  0.333333  0.2500   
139  18684    2021  Sandy Alcantara  MIA  0.222222  0.166667  0.7500   
164  19427    2019     Shane Bieber  CLE  0.166667  0.500000  0.3125   

          WAR       ERA      G  ...   HardHit    Events     CStr%      CSW%  \
241  0.352273  0.559659  0.250  ...  0.000000  0.000000  0.391892  0.261538   
286  0.454545  0.605114  0.750  ...  0.000000  0.000000  0.391892  0.176923   
295  0.318182  0.619318  0.625  ...  0.000000  0.000000  0.459459  0.200000   
353  0.409091  0.744318  0.625  ...  0.000000  0.000000  0.527027  0.200000   
237  0.329545  0.548295  0.375  ...  0.000000  0.000000  0.500000  0.215385   
..        ...       ...    ...  ...       ...       ...       ...       ...   
360  0.352273  0.769886  0.500  ...  1.000000  0.919825  0.405405  0.369231   
13   0.602273  0.201705  0.625  ...  0.636364  0.790087  0.756757  0.815385   
301  0.363636  0.627841  0.750  ...  0.795455  0.775510  0.864865  0.761538   
139  0.465909  0.434659  0.625  ...  0.859848  0.839650  0.054054  0.484615   
164  0.602273  0.460227  0.750  ...  0.875000  0.807580  0.581081  0.838462   

     Next_WAR  Team_code  player_season  war_corr  war_diff  war_season  
241       4.2    0.18750              0  0.000000  1.000000    0.746239  
286       3.0    0.18750              1  1.000000  1.290323    0.904419  
295       3.8    0.18750              2 -0.240192  0.700000    0.648749  
353       3.1    0.18750              3  0.072868  1.285714    0.989176  
237       1.6    0.18750              4 -0.249513  0.805556    0.711482  
..        ...        ...            ...       ...       ...         ...  
360       3.4    0.96875              0  0.000000  1.000000    0.750000  
13        3.4    0.68750              0  0.000000  1.000000    0.966146  
301       6.3    0.68750              1 -1.000000  0.603774    0.607595  
139       5.7    0.50000              0  0.000000  1.000000    1.000000  
164       4.9    0.25000              0  0.000000  1.000000    1.006329  

[235 rows x 158 columns]

In [45]:
# Add more columns to predictors
new_predictors = predictors + ["player_season", "war_corr", "war_season", "war_diff"]
pitching[new_predictors]

L       WAR      G       IBB       IFH      WHIP      LOB%  \
241  0.3750  0.352273  0.250  0.555556  0.620690  0.552632  0.490566   
286  0.5625  0.454545  0.750  0.444444  0.793103  0.657895  0.306604   
295  0.4375  0.318182  0.625  0.333333  0.379310  0.539474  0.443396   
353  0.6250  0.409091  0.625  0.111111  0.413793  0.736842  0.231132   
237  0.3750  0.329545  0.375  0.333333  0.620690  0.605263  0.377358   
..      ...       ...    ...       ...       ...       ...       ...   
360  0.4375  0.352273  0.500  0.000000  0.448276  0.592105  0.155660   
13   0.1875  0.602273  0.625  0.333333  0.517241  0.171053  0.811321   
301  0.2500  0.363636  0.750  0.333333  0.241379  0.565789  0.542453   
139  0.7500  0.465909  0.625  0.222222  0.620690  0.302632  0.334906   
164  0.3125  0.602273  0.750  0.111111  0.448276  0.276316  0.570755   

         inLI     FB% 2      FBv  ...      RS/9  Z-Contact% (sc)  BIP-Wins  \
241  0.458333  0.451034  0.25000  ...  0.253863         0.704301  0.523077   
286  0.250000  0.499310  0.28750  ...  0.551876         0.892473  0.276923   
295  0.416667  0.445517  0.23750  ...  0.293598         0.876344  0.538462   
353  0.250000  0.569655  0.25625  ...  0.410596         0.946237  0.200000   
237  0.541667  0.459310  0.23125  ...  0.273731         0.844086  0.384615   
..        ...       ...      ...  ...       ...              ...       ...   
360  0.541667  0.628966  0.65625  ...  0.328918         0.779570  0.307692   
13   0.791667  0.520000  0.65625  ...  0.167770         0.370968  0.661538   
301  0.500000  0.473103  0.68750  ...  0.543046         0.596774  0.384615   
139  0.791667  0.525517  1.00000  ...  0.019868         0.489247  0.523077   
164  0.625000  0.464828  0.70000  ...  0.428256         0.569892  0.415385   

        Hard%  LD%+    Cent%+  player_season  war_corr  war_season  war_diff  
241  0.483193  0.50  0.264706              0  0.000000    0.746239  1.000000  
286  0.239496  0.46  0.529412              1  1.000000    0.904419  1.290323  
295  0.126050  0.54  0.352941              2 -0.240192    0.648749  0.700000  
353  0.294118  0.30  0.294118              3  0.072868    0.989176  1.285714  
237  0.210084  0.56  0.735294              4 -0.249513    0.711482  0.805556  
..        ...   ...       ...            ...       ...         ...       ...  
360  0.542017  0.44  0.500000              0  0.000000    0.750000  1.000000  
13   0.264706  0.28  0.529412              0  0.000000    0.966146  1.000000  
301  0.970588  0.44  0.588235              1 -1.000000    0.607595  0.603774  
139  0.441176  0.34  0.470588              0  0.000000    1.000000  1.000000  
164  1.000000  0.50  0.588235              0  0.000000    1.006329  1.000000  

[235 rows x 24 columns]

In [46]:
predictions = backtest(pitching, rr, new_predictors)
predictions

actual  prediction
286     3.0    3.594672
46      5.8    5.722762
333     2.8    4.107397
96      3.6    3.927717
229     4.4    2.970836
..      ...         ...
13      3.4    4.521565
30      3.3    7.184415
301     6.3    3.871178
164     4.9    4.681119
139     5.7    4.876636

[211 rows x 2 columns]

### Calculate new accuracy 

In [47]:
# Calculate the mean squared error value (To compare the prediction from actual value)
mean_squared_error(predictions["actual"], predictions["prediction"]) 

2.4633736374384574

Value slighly bigger than before. First predictors gave better accurancy. 

In [49]:
# Sort predictors coefficient to check which columns contributed more to the model 
pd.Series(rr.coef_, index=new_predictors).sort_values()

G                 -1.130219
Cent%+            -0.708083
LD%+              -0.590452
Z-Contact%        -0.588266
inLI              -0.583023
Z-Contact% (sc)   -0.512338
BIP-Wins          -0.473151
IFH               -0.438200
war_corr          -0.141806
war_diff          -0.087096
RS/9              -0.073346
FB% 2             -0.058328
player_season      0.006613
F-Strike%          0.502936
IBB                0.617915
L                  0.672567
wFB/C              0.751776
LOB%               0.908267
Hard%              0.932547
WHIP               1.419711
SD                 1.428966
WAR                1.624767
war_season         1.929515
FBv                1.977000
dtype: float64

All small values indicated that the model is not taking these columns into account to make the prediction.
Let's try to remove columns with correlation less than 30$


In [52]:
final_predictors=['L',
 'WAR',
 'G',
 'IBB',
 'IFH',
 'WHIP',
 'LOB%',
 'inLI',
 'FBv',
 'wFB/C',
 'Z-Contact%',
 'F-Strike%',
 'SD',
 'Z-Contact% (sc)',
 'BIP-Wins',
 'Hard%',
 'LD%+',
 'Cent%+', 
 'war_season',
 ]

In [54]:
final_prediction = backtest(pitching, rr, final_predictors)
final_prediction

actual  prediction
286     3.0    3.527590
46      5.8    5.608402
333     2.8    4.148904
96      3.6    3.894350
229     4.4    3.024789
..      ...         ...
13      3.4    4.454356
30      3.3    7.132866
301     6.3    3.821859
164     4.9    4.694646
139     5.7    4.848092

[211 rows x 2 columns]

### Calculate last accuracy 

In [55]:
mean_squared_error(final_predictions["actual"], final_predictions["prediction"]) 

2.29445344177085

This mean squared error is the smallest meaning the last model is the most accurate 

### Check the difference between predicted and actual WAR nb 

In [56]:
merged = final_prediction.merge(pitching, left_index=True, right_index=True)
merged["diff"] = (final_prediction["actual"] - final_prediction["prediction"]).abs()
next_war_df=merged[["IDfg", "Season", "Name", "WAR", "Next_WAR", "diff","prediction"]].sort_values(["diff"])
next_war_df.head()

IDfg  Season               Name       WAR  Next_WAR      diff  \
24    4930    2014         Jon Lester  0.590909       4.9  0.032192   
175  11423    2014      Jose Quintana  0.534091       4.3  0.037476   
118   4972    2008        Cole Hamels  0.500000       3.6  0.038084   
89    5524    2015  Madison Bumgarner  0.534091       4.3  0.072897   
153   7448    2010       Gio Gonzalez  0.340909       3.3  0.080982   

     prediction  
24     4.932192  
175    4.262524  
118    3.561916  
89     4.227103  
153    3.219018

In [57]:
next_war_df = next_war_df.groupby("IDfg", group_keys=False).apply(next_season)
next_war_df

IDfg  Season             Name       WAR  Next_WAR      diff  prediction
286    225    2008     Mark Buehrle  0.454545  0.318182  0.527590    3.527590
295    225    2009     Mark Buehrle  0.318182  0.409091  0.399684    3.400316
353    225    2010     Mark Buehrle  0.409091  0.329545  1.293168    4.393168
237    225    2011     Mark Buehrle  0.329545  0.159091  1.162463    2.762463
271    225    2012     Mark Buehrle  0.159091  0.227273  0.403668    1.796332
..     ...     ...              ...       ...       ...       ...         ...
360  13431    2016   Marcus Stroman  0.352273       NaN  0.239558    3.160442
13   16149    2018       Aaron Nola  0.602273  0.363636  1.054356    4.454356
301  16149    2019       Aaron Nola  0.363636       NaN  2.478141    3.821859
139  18684    2021  Sandy Alcantara  0.465909       NaN  0.851908    4.848092
164  19427    2019     Shane Bieber  0.602273       NaN  0.205354    4.694646

[211 rows x 7 columns]

In [58]:
import json
next_war_js=next_war_df.to_json(orient = "records")
next_war_js

'[{"IDfg":225,"Season":2008,"Name":"Mark Buehrle","WAR":0.4545454545,"Next_WAR":0.3181818182,"diff":0.5275900903,"prediction":3.5275900903},{"IDfg":225,"Season":2009,"Name":"Mark Buehrle","WAR":0.3181818182,"Next_WAR":0.4090909091,"diff":0.3996835393,"prediction":3.4003164607},{"IDfg":225,"Season":2010,"Name":"Mark Buehrle","WAR":0.4090909091,"Next_WAR":0.3295454545,"diff":1.293167659,"prediction":4.393167659},{"IDfg":225,"Season":2011,"Name":"Mark Buehrle","WAR":0.3295454545,"Next_WAR":0.1590909091,"diff":1.1624631814,"prediction":2.7624631814},{"IDfg":225,"Season":2012,"Name":"Mark Buehrle","WAR":0.1590909091,"Next_WAR":0.2272727273,"diff":0.4036675629,"prediction":1.7963324371},{"IDfg":225,"Season":2013,"Name":"Mark Buehrle","WAR":0.2272727273,"Next_WAR":null,"diff":1.5027367247,"prediction":2.1972632753},{"IDfg":232,"Season":2009,"Name":"Jon Garland","WAR":0.1818181818,"Next_WAR":null,"diff":2.7283778314,"prediction":3.6283778314},{"IDfg":404,"Season":2008,"Name":"CC Sabathia","WAR

In [59]:
# Save JSON file in clean data file
# save_file=open('../static/data/Clean_Data/next_war.js','w')
json.dump(next_war_js, save_file, indent = 6)  
save_file.close() 

# Pitching Stats & Records Data

In [ ]:
# Add an option on website for user to check the records of previous games 
data_test = schedule_and_record(2017, 'NYY')
data_test.head()